### PR TITLE
Make sttm-desktop compatible with RPi

### PR DESCRIPTION
**Pull request by Gauravjeetsingh**
- Updates electron to version 19 with all the breaking changes implemented.
- Updates realm to version 10
- Uses node 16
- Replaces openport package with portfinder
- Replaces react-html-parser with html-react-parser package

and other minor package version upgrades.

### CODE REVIEW

**Comment by SikhNerd** [comment]
I installed a fresh, latest version of rpiOS on my pi400 and checked out this branch.

`npm ci` completed correctly, however running `npm run pack:linux` gives the following error

```
  $ npm run pack:linux

> sttm-desktop@9.0.4 pack:linux
> npm run build && electron-builder --linux --x64


> sttm-desktop@9.0.4 build
> npm run build-css & npm run build-js


> sttm-desktop@9.0.4 build-css
> run-p build-css:*


> sttm-desktop@9.0.4 build-js
> rm -rf www/js & babel ./www/main -d ./www/js


> sttm-desktop@9.0.4 build-css:main
> node-sass --output-style compressed --include-path www/src/scss www/src/scss/styles.scss www/assets/css/bundle.css


> sttm-desktop@9.0.4 build-css:obs
> node-sass --output-style compressed --include-path www/src/scss/obs www/src/scss/obs/styles.scss www/obs/bundle.css

Rendering Complete, saving .css file...
Wrote CSS to /home/pi/s/sttm-desktop/www/obs/bundle.css
Rendering Complete, saving .css file...
Wrote CSS to /home/pi/s/sttm-desktop/www/assets/css/bundle.css
Successfully compiled 203 files with Babel (14320ms).
  • electron-builder  version=23.1.0 os=5.15.32-v7l+
  • loaded configuration  file=package.json ("build" field)
  • electron-rebuild not required if you use electron-builder, please consider to remove excess dependency from devDependencies

To ensure your native dependencies are always matched electron version, simply add script `"postinstall": "electron-builder install-app-deps" to your `package.json`
  • writing effective config  file=builds/builder-effective-config.yaml
  • rebuilding native dependencies  dependencies=mdns@2.7.2 platform=linux arch=x64

  ⨯ cannot execute  cause=exit status 1
                    errorOut=npm ERR! code 1
    npm ERR! path /home/pi/s/sttm-desktop/node_modules/mdns
    npm ERR! command failed
    npm ERR! command sh -c node-gyp rebuild
    npm ERR! make: Entering directory '/home/pi/s/sttm-desktop/node_modules/mdns/build'
    npm ERR!   CXX(target) Release/obj.target/dns_sd_bindings/src/dns_sd.o
    npm ERR! make: Leaving directory '/home/pi/s/sttm-desktop/node_modules/mdns/build'
    npm ERR! gyp info it worked if it ends with ok
    npm ERR! gyp info using node-gyp@9.1.0
    npm ERR! gyp info using node@16.16.0 | linux | arm
    npm ERR! gyp info find Python using Python version 3.9.2 found at "/usr/bin/python3"
    npm ERR! gyp info spawn /usr/bin/python3
    npm ERR! gyp info spawn args [
    npm ERR! gyp info spawn args   '/home/pi/s/sttm-desktop/node_modules/node-gyp/gyp/gyp_main.py',
    npm ERR! gyp info spawn args   'binding.gyp',
    npm ERR! gyp info spawn args   '-f',
    npm ERR! gyp info spawn args   'make',
    npm ERR! gyp info spawn args   '-I',
    npm ERR! gyp info spawn args   '/home/pi/s/sttm-desktop/node_modules/mdns/build/config.gypi',
    npm ERR! gyp info spawn args   '-I',
    npm ERR! gyp info spawn args   '/home/pi/s/sttm-desktop/node_modules/node-gyp/addon.gypi',
    npm ERR! gyp info spawn args   '-I',
    npm ERR! gyp info spawn args   '/home/pi/.electron-gyp/19.0.8/include/node/common.gypi',
    npm ERR! gyp info spawn args   '-Dlibrary=shared_library',
    npm ERR! gyp info spawn args   '-Dvisibility=default',
    npm ERR! gyp info spawn args   '-Dnode_root_dir=/home/pi/.electron-gyp/19.0.8',
    npm ERR! gyp info spawn args   '-Dnode_gyp_dir=/home/pi/s/sttm-desktop/node_modules/node-gyp',
    npm ERR! gyp info spawn args   '-Dnode_lib_file=/home/pi/.electron-gyp/19.0.8/<(target_arch)/node.lib',
    npm ERR! gyp info spawn args   '-Dmodule_root_dir=/home/pi/s/sttm-desktop/node_modules/mdns',
    npm ERR! gyp info spawn args   '-Dnode_engine=v8',
    npm ERR! gyp info spawn args   '--depth=.',
    npm ERR! gyp info spawn args   '--no-parallel',
    npm ERR! gyp info spawn args   '--generator-output',
    npm ERR! gyp info spawn args   'build',
    npm ERR! gyp info spawn args   '-Goutput_dir=.'
    npm ERR! gyp info spawn args ]
    npm ERR! gyp info spawn make
    npm ERR! gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
    npm ERR! g++: error: unrecognized command-line option ‘-m64’
    npm ERR! make: *** [dns_sd_bindings.target.mk:198: Release/obj.target/dns_sd_bindings/src/dns_sd.o] Error 1
    npm ERR! gyp ERR! build error 
    npm ERR! gyp ERR! stack Error: `make` failed with exit code: 2
    npm ERR! gyp ERR! stack     at ChildProcess.onExit (/home/pi/s/sttm-desktop/node_modules/node-gyp/lib/build.js:201:23)
    npm ERR! gyp ERR! stack     at ChildProcess.emit (node:events:527:28)
    npm ERR! gyp ERR! stack     at Process.ChildProcess._handle.onexit (node:internal/child_process:291:12)
    npm ERR! gyp ERR! System Linux 5.15.32-v7l+
    npm ERR! gyp ERR! command "/usr/bin/node" "/home/pi/s/sttm-desktop/node_modules/.bin/node-gyp" "rebuild"
    npm ERR! gyp ERR! cwd /home/pi/s/sttm-desktop/node_modules/mdns
    npm ERR! gyp ERR! node -v v16.16.0
    npm ERR! gyp ERR! node-gyp -v v9.1.0
    npm ERR! gyp ERR! not ok
    
    npm ERR! A complete log of this run can be found in:
    npm ERR!     /home/pi/.npm/_logs/2022-08-15T23_32_11_425Z-debug-0.log
    
                    command=/usr/bin/node /usr/lib/node_modules/npm/bin/npm-cli.js rebuild mdns@2.7.2
                    workingDir=
```

System info 
```
pi@raspberrypi:~/s/sttm-desktop $ node -v
v16.16.0
pi@raspberrypi:~/s/sttm-desktop $ npm version
{
  'sttm-desktop': '9.0.4',
  npm: '8.11.0',
  node: '16.16.0',
  v8: '9.4.146.24-node.21',
  uv: '1.43.0',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.18.1',
  modules: '93',
  nghttp2: '1.47.0',
  napi: '8',
  llhttp: '6.0.7',
  openssl: '1.1.1q+quic',
  cldr: '40.0',
  icu: '70.1',
  tz: '2021a3',
  unicode: '14.0',
  ngtcp2: '0.1.0-DEV',
  nghttp3: '0.1.0-DEV'
}
pi@raspberrypi:~/s/sttm-desktop $ lsb_release -a
No LSB modules are available.
Distributor ID:	Raspbian
Description:	Raspbian GNU/Linux 11 (bullseye)
Release:	11
Codename:	bullseye
pi@raspberrypi:~/s/sttm-desktop $ uname -a
Linux raspberrypi 5.15.32-v7l+ #1538 SMP Thu Mar 31 19:39:41 BST 2022 armv7l GNU/Linux
```

**Comment by SikhNerd** [comment]
`npm run start` completes, but gives no GUI

**Comment by SikhNerd** [comment]
I realize now that pack:linux is specifically for x64, I'll come back to this by trying to focus on the `npm run start` issue

**Comment by Gauravjeetsingh** [comment]
Yeah, `npm run pack:linux` won't work since that's for x64 build. I am still figuring out how to create a build for RPi.

I currently tested this on ubuntu 22.04 in raspberry pi. It works perfectly fine there. Do you see any error in browser console or the terminal after you run `npm start`?

My system info
```
paxi@berry:~$ npm version
{
  npm: '8.1.2',
  node: '16.13.1',
  v8: '9.4.146.24-node.14',
  uv: '1.42.0',
  zlib: '1.2.11',
  brotli: '1.0.9',
  ares: '1.18.1',
  modules: '93',
  nghttp2: '1.45.1',
  napi: '8',
  llhttp: '6.0.4',
  openssl: '1.1.1l+quic',
  cldr: '39.0',
  icu: '69.1',
  tz: '2021a',
  unicode: '13.0',
  ngtcp2: '0.1.0-DEV',
  nghttp3: '0.1.0-DEV'
}
paxi@berry:~$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04 LTS
Release:	22.04
Codename:	jammy
paxi@berry:~$ uname -a
Linux berry 5.15.0-1012-raspi #14-Ubuntu SMP PREEMPT Fri Jun 24 13:10:28 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux
paxi@berry:~$ 

```

File: www/main/viewer/utils/chromecast.js [hunk 1]
```diff
@@ -0,0 +1,182 @@
...
+	sendMessage(JSON.stringify(getSanitizedViewer()));
+};
+
+/**
+ * append message to debug message window
+ * @param {string} message A message string
+ */
+export const appendMessage = (message) => {
+	console.log(message);
+}
```
Comment (theamanjs) [inline www/main/viewer/utils/chromecast.js:182]: Please add new blank line at the eof

File: www/main/viewer/utils/index.js [hunk 2]
```diff
@@ -1 +1,2 @@
 export { updateViewerScale } from './ScaleViewer';
+export { castToReceiver, appendMessage, requestSession, stopApp, tingle } from './chromecast';
```
Comment (theamanjs) [inline www/main/viewer/utils/index.js:2]: Please add new blank line at the eof

**Review by theamanjs — CHANGES_REQUESTED** [review]

File: www/main/viewer/utils/index.js (hunk 2 above)
Comment (Gauravjeetsingh) [reply www/main/viewer/utils/index.js]: done.

File: www/main/viewer/utils/chromecast.js (hunk 1 above)
Comment (Gauravjeetsingh) [reply www/main/viewer/utils/chromecast.js]: done.

File: www/main/controller.js [hunk 3]
```diff
@@ -447,92 +442,20 @@ global.platform.ipc.on('cast-session-stopped', () => {
...
 
 global.platform.ipc.on('set-user-setting', (event, settingChanger) => {
-  if (settingChanger.func === 'size') {
-    changeFontSize(settingChanger.iconType, settingChanger.operation === 'plus');
-  } else if (settingChanger.func === 'visibility') {
-    changeVisibility(settingChanger.iconType);
+  const { func, iconType, operation } = JSON.parse(settingChanger);
+  if (func === 'size') {
+    changeFontSize(iconType, operation === 'plus');
+  } else if (func === 'visibility') {
```
[the remark this answers is not in the document]
Comment (Gauravjeetsingh) [reply www/main/controller.js:447]: Thank you ji! :)

File: www/main/controller.js [hunk 4]
```diff
@@ -370,7 +363,7 @@ function checkPresenterView() {
   classList.toggle('scale-viewer', inPresenterView);
 
   global.platform.ipc.send('presenter-view', inPresenterView);
-  global.webview.send('presenter-view', inPresenterView);
+  // global.webview.send('presenter-view', inPresenterView);
```
[the remark this answers is not in the document]
Comment (Gauravjeetsingh) [reply www/main/controller.js]: Yes, removed.

File: www/main/banidb/realm-search.js [hunk 5]
```diff
@@ -160,14 +167,20 @@ const query = (searchQuery, searchType, searchSource) =>
       default:
         break;
     }
-    order.push('Shabads');
-    condition = `${condition} SORT(${order.join(' ASC, ')} ASC)`;
+    //  order.push('Shabads');
```
[the remark this answers is not in the document]
Comment (Gauravjeetsingh) [reply www/main/banidb/realm-search.js]: Done.

File: www/main/addons/bani-controller/utils/share-sync.js [hunk 6]
```diff
@@ -33,7 +35,7 @@ async function getNewCode(host) {
 
     newCode = namespaceString;
   } catch (error) {
-    analytics.trackEvent('sync', 'error', error);
+    // analytics.trackEvent('sync', 'error', error);
```
[the remark this answers is not in the document]
Comment (Gauravjeetsingh) [reply www/main/addons/bani-controller/utils/share-sync.js]: Commented while testing something, uncommented and pushed.

**Review by theamanjs — DISMISSED** [review]

**Comment by theamanjs** [comment]
Looks good to me veerji @Gauravjeetsingh :taco:

**Comment by theamanjs** [comment]
Just a reminder veerji, we still don't have any changelog attached with this.

**Comment by Gauravjeetsingh** [comment]
@theamanjs ji, its okay. We can add the changelog text in a separate PR. Can you please re-review this again, added some missing commits.

cc: @inderpreetsingh

File: www/main/desktop_scripts.js [hunk 7]
```diff
@@ -127,50 +127,50 @@ module.exports = {
...
+                        const win = remote.getCurrentWindow();
+                        win.setProgressBar(-1); 
                       });
-                      const win = remote.getCurrentWindow();
-                      win.setProgressBar(-1);
-                    });
+                    } catch (err) {
+                      // handle any errors
+                      console.log(err0);
+                    }
```
Comment (theamanjs) [inline www/main/desktop_scripts.js:173]: Veerji, aren't we gonna show this error on frontend?

File: www/src/scss/viewer/verse-slide/_verse-slide.scss [hunk 8]
```diff
@@ -51,6 +51,14 @@
...
   margin: 10px 0;
   order: 2;
+	&.custom-english {
+		display: flex;
+		flex-direction: column;
+		flex-wrap: wrap;
+		height: 100%;
+		overflow-y: auto;
+		justify-content: space-evenly;
+	}
```
Comment (theamanjs) [inline www/src/scss/viewer/verse-slide/_verse-slide.scss:61]: Veerji looks like some indentation issues are there

**Review by theamanjs — CHANGES_REQUESTED** [review]

File: www/src/scss/viewer/verse-slide/_verse-slide.scss (hunk 8 above)
Comment (theamanjs) [reply www/src/scss/viewer/verse-slide/_verse-slide.scss]: Fixed here https://github.com/KhalisFoundation/sttm-desktop/pull/1628/commits/228548c0479034884b8753b8e3ee88935ca0e3e2

File: www/main/desktop_scripts.js (hunk 7 above)
Comment (Gauravjeetsingh) [reply www/main/desktop_scripts.js:173]: This is from old code, lets keep it as it is for now and we can have a separate PR for this later.

Comment (theamanjs) [reply www/main/desktop_scripts.js:173]: Sure veerji thanks

**Review by theamanjs — APPROVED** [review]

**Pull request merged**